util/fetch/src/client.rs

```diff
@@ -14,347 +14,554 @@
-			Ok(client) => {
-				self.pool.spawn(FetchTask {
-					url: url.into(),
-					client: client,
-					limit: self.limit,
-					abort: abort,
-				})
-			},
-			Err(err) => {
-				self.pool.spawn(futures::future::err(err))
-			},
+	/// Is the content-type text/html?
+	pub fn is_html(&self) -> bool {
+		if let Some(ref mime) = self.content_type() {
+			mime.type_() == "text" && mime.subtype() == "html"
+		} else {
+			false
 		}
 	}
-}
 
-struct FetchTask {
-	url: String,
-	client: Arc<reqwest::Client>,
-	limit: Option<usize>,
-	abort: Abort,
+	/// The conten-type header value.
+	pub fn content_type(&self) -> Option<Mime> {
+		self.headers.get::<ContentType>().map(|ct| ct.0.clone())
+	}
 }
 
-impl Future for FetchTask {
-	// TODO [ToDr] timeouts handling?
-	type Item = Response;
+impl Stream for Response {
+	type Item = hyper::Chunk;
 	type Error = Error;
 
-	fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+	fn poll(&mut self) -> futures::Poll<Option<Self::Item>, Self::Error> {
 		if self.abort.is_aborted() {
-			trace!(target: "fetch", "Fetch of {:?} aborted.", self.url);
-			return Err(Error::Aborted);
+			debug!(target: "fetch", "fetch of {} aborted", self.url);
+			return Err(Error::Aborted)
 		}
+		match try_ready!(self.body.poll()) {
+			None    => Ok(Async::Ready(None)),
+			Some(c) => {
+				if self.nread + c.len() > self.abort.max_size() {
+					debug!(target: "fetch", "size limit {:?} for {} exceeded", self.abort.max_size(), self.url);
+					return Err(Error::SizeLimit)
+				}
+				self.nread += c.len();
+				Ok(Async::Ready(Some(c)))
+			}
+		}
+	}
+}
 
-		trace!(target: "fetch", "Starting fetch task: {:?}", self.url);
-		let result = self.client.get(&self.url)
-						  .header(reqwest::header::UserAgent::new("Parity Fetch"))
-						  .send()?;
+/// `BodyReader` serves as an adapter from async to sync I/O.
+///
+/// It implements `io::Read` by repedately waiting for the next `Chunk`
+/// of hyper's response `Body` which blocks the current thread.
+pub struct BodyReader {
+	chunk:  hyper::Chunk,
+	body:   Option<hyper::Body>,
+	offset: usize,
+}
 
-		Ok(futures::Async::Ready(Response {
-			inner: ResponseInner::Response(result),
-			abort: self.abort.clone(),
-			limit: self.limit,
-			read: 0,
-		}))
+impl BodyReader {
+	/// Create a new body reader for the given response.
+	pub fn new(r: Response) -> BodyReader {
+		BodyReader {
+			body: Some(r.body),
+			chunk: Default::default(),
+			offset: 0,
+		}
 	}
 }
 
-/// Fetch Error
+impl io::Read for BodyReader {
+	fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+		let mut m = 0;
+		while self.body.is_some() {
+			// Can we still read from the current chunk?
+			if self.offset < self.chunk.len() {
+				let k = cmp::min(self.chunk.len() - self.offset, buf.len() - m);
+				let c = &self.chunk[self.offset .. self.offset + k];
+				(&mut buf[m .. m + k]).copy_from_slice(c);
+				self.offset += k;
+				m += k;
+				if m == buf.len() {
+					break
```

Comment:
I would put `return Ok(m)` here and instead of the other `break` - thanks to this we don't need the `Ok(m)` at the bottom

Reply:
I think `Ok(m)` is still needed there, as the `while` loop might theoretically not be entered at all.

Reply:
Oh, that's true. We could use `loop` and `if let Some(body) = self.body.take()` inside to avoid `expect`?